### PR TITLE
RD2 Custom Form: Donor Type Selection

### PR DESCRIPTION
**Pull request by force2b**
- Separate the Donor section of the RD2 UI form into a child component
- Add a picklist for "Donor Type" to select the type of Recurring Donation to create.

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

Lightning Web Components:
- rd2EntryFormDonorSection

# Deleted Metadata

### CODE REVIEW

File: src/labels/CustomLabels.labels [hunk 1]
```diff
@@ -1983,6 +1983,22 @@
         <shortDescription>Recurring Donation Entry Form Donor Information section header</shortDescription>
         <value>Donor Information</value>
     </labels>
+    <labels>
+        <fullName>RD2_EntryFormDonorTypeHelpText</fullName>
+        <categories>Recurring-Donations, Entry-Form</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>The Inline HelpText for the Donor Type picklist</shortDescription>
+        <value>Select the type of donor for this recurring donation. A value of &quot;Account&quot; requires entry of an Organization Account and optionally a related Contact. A value of &quot;Contact&quot; will require entry of a Household Contact.</value>
```
Comment (force2b) [inline src/labels/CustomLabels.labels:1992]: @sam-knox Would you mind reviewing the help text for the "Donor Type" field on the RD New/Edit UI?

**Comment by force2b** [comment]
**lurch: attach W-040321
Also W-034369

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HttEQAS" target="_blank">W-040321</a>

File: src/lwc/geListView/geListView.js [hunk 2]
```diff
@@ -5,13 +5,13 @@ import {
...
     handleError,
     generateId,
-    showToast,
 } from 'c/utilTemplateBuilder'
 import {
     format,
     deepClone,
     isNotEmpty,
     hasNestedProperty,
+    showToast,
```
Comment (force2b) [inline src/lwc/geListView/geListView.js:14]: I worked with the HoneyGems to discuss moving showToast to the utilCommon component

File: src/lwc/utilCommon/utilCommon.js [hunk 3]
```diff
@@ -429,7 +433,76 @@ const validateJSONString = (str) => {
...
+ *
+ * @param {string} title: Title of the toast, displayed as a heading.
+ * @param {string} message: Message of the toast. It can contain placeholders in
+ * the form of {0} ... {N}. The placeholders are replaced with the links from
+ * messageData param
+ * @param {string} mode: Mode of the toast
+ * @param {array} messageData: List of values that replace the {index} placeholders
+ * in the message param
+ */
+const showToast = (title, message, variant, mode, messageData) => {
```
Comment (force2b) [inline src/lwc/utilCommon/utilCommon.js:493]: I worked with the HoneyGems to discuss moving showToast to the utilCommon component

**Comment by npsp-reedestockton** [comment]
cci flow enable_rd2 fails:

`2020-06-12 15:58:07: [Failed]: Update of CustomObject npe03__Recurring_Donation__c: Error: RD2_EntryForm does not exist or is not a valid override for action New.
2020-06-12 15:58:07: Exception in task deploy_rd2_config
`

File: src/lwc/rd2EntryFormDonorSection/rd2EntryFormDonorSection.html [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+                <lightning-layout horizontal-align="spread" multiple-rows="true">
+
+                    <lightning-layout-item padding="horizontal-small" size="12" small-device-size="6">
+                        <lightning-combobox
+                                name="DonorTypeSelect"
+                                label={customLabels.donorTypeLabel}
+                                value={donorType}
+                                placeholder=""
+                                field-level-help={customLabels.donorTypeHelpText}
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/lwc/rd2EntryFormDonorSection/rd2EntryFormDonorSection.html:17]: Good catch. Fixed.

**Comment by force2b** [comment]
@npsp-reedestockton There's actually no code in the component to do anything with the Account or Contact selected, so I'm not sure why selecting an Org Contact would anything different on this page than it would elsewhere. What did you select for "Donor Type" when you got this error?

**Comment by force2b** [comment]
@npsp-reedestockton I am able to reproduce your error though. Looking into it.

File: src/lwc/rd2EntryForm/rd2EntryForm.js [hunk 5]
```diff
@@ -206,13 +185,18 @@ export default class rd2EntryForm extends LightningElement {
...
+        const scheduleSection = this.scheduleComponent;
         const scheduleFields = (scheduleSection === null || scheduleSection === undefined)
             ? {}
             : scheduleSection.returnValues();
 
+        const donorSection = this.donorComponent;
+        const donorFields = (donorSection === null || donorSection === undefined)
+            ? {}
+            : donorSection.returnValues();
+
```
Comment (andrewyu-salesforce) [inline src/lwc/rd2EntryForm/rd2EntryForm.js:197]: You can add a let isValid = donorSection.validateFields(); in here to check the required field missing validation, and add the following function in the rd2EntryFormDonorSection.js 
` @api
    validateFields() {
    let isValid = true;
    this.template.querySelectorAll('lightning-input-field')
            .forEach(field => {
                isValid = field.reportValidity()
            });
        return isValid;
    }`

Then only submit the form if isValid = true

**Comment by sam-knox** [comment]
@andrewyu-salesforce - Instead of "Complete this field" for the validation error message, can we use "Contact is a required field."?

**Comment by andrewyu-salesforce** [comment]
@sam-knox The 'Complete this field' validation can be customize to anything. However, this is the standard Salesforce validation message for required field. If we want to change the contact's validation message, I think it will be better to also customize all other required fields' messages on the entry form. @force2b

File: src/lwc/rd2EntryForm/rd2EntryForm.js (hunk 5 above)
Comment (force2b) [reply src/lwc/rd2EntryForm/rd2EntryForm.js:197]: Thanks @andrewyu-salesforce. Changes made.

**Comment by force2b** [comment]
@sam-knox and @andrewyu-salesforce It doesn't appear that the built-in validation messages (ex: "Complete this Field") can be customized in LWC. I tried a number of variations. Even the documented `setCustomValidity()` javascript function isn't available in LWC. For now I think we're stuck with the message that Salesforce generates for these.

File: src/labels/CustomLabels.labels (hunk 1 above)
Comment (force2b) [reply src/labels/CustomLabels.labels:1992]: @sam-knox Just checking back on this new label.

**Comment by andrewyu-salesforce** [comment]
@force2b Everything looks great

File: src/labels/CustomLabels.labels (hunk 1 above)
Comment (sam-knox) [reply src/labels/CustomLabels.labels:1992]: Yes, sorry I missed this. Here's my suggested update to the Value:

"Select the type of donor for this Recurring Donation. If you select Account, enter an Organization Account, and optionally a related Contact. If you select Contact, enter a Contact related to a Household Account."

Comment (force2b) [reply src/labels/CustomLabels.labels]: Thanks @sam-knox. Change made.

**Comment by rponti-sforg** [comment]
@force2b @andrewyu-salesforce Good morning.  Wanted to let you know that I am all finished testing the WIs associated with this PR.  Testing passed.  I restarted the Feature Test - Robot last night to see if we could get a pass on that.  The Namespaced one (I think it was that) was still running.  It looks like Mike restarted that this morning due to it failing.  I will keep my eye on this but will be stepping away to run some errands.

**Comment by rponti-sforg** [comment]
QE testing is now complete for this PR and associated WIs.  Based on the WI's Descriptions, ACs, UI design documents, test charters, and additional feedback, tests have passed.

For further WI and testing details, see:
[W-034369](https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001C7TKQA0/view)
https://salesforce.quip.com/zYKMAcbD0snD
[W-040185](https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001HtclQAC/view)
https://salesforce.quip.com/xLJDA8moOYZw
[W-040321](https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001HttEQAS/view)
https://salesforce.quip.com/7VXFA6P5pQu3
[W-041485](https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000007haNEQAY/view)

WI statuses will be set to 'Closed' once post merge builds have successfully passed.

**Pull request merged**